### ISSUE TITLE
The query is very slow when some nodes of the redis cluster are unavailable

### ISSUE THREAD

**Issue by liuzhaozhao** [quarkusio/quarkus#20159]
### Describe the bug

application.yml
The first three addresses in the configuration "quarkus.redis.hosts" are unavailable and do not belong to the cluster node (redis cluster shrinkage)
```yml
quarkus:
  redis:
    health:
      enabled: true
    hosts: redis://127.0.0.1:5138,redis://127.0.0.1:5139,redis://127.0.0.1:5140,redis://127.0.0.1:5141,redis://127.0.0.1:7017,redis://127.0.0.1:7018,redis://127.0.0.1:7019,redis://127.0.0.1:7020,redis://127.0.0.1:7021
    password: xxxxxxxxx
    timeout: 10s
    client-type: cluster
    replicas: share
    max-pool-size: 20
    max-pool-waiting: 24
    pool-cleaner-interval: 5s
    pool-recycle-timeout: 15s
    max-waiting-handlers: 2048
    max-nested-arrays: 16
    reconnect-interval: 30s
```

code
```java
@Tag(name = "redis sample")
@Slf4j
@Path("/redis")
public class RedisResource {

    @Inject
    RedisService redisService;

    @Operation(summary = "get redis key value")
    @GET @Path("/{key}") @Produces(MediaType.APPLICATION_JSON)
    @Blocking
    public Response<String> get(@PathParam("key") String key) {
        return Response.ok(redisService.get(key));
    }
}
```
```java
@Singleton
public class RedisService {
    @Inject
    RedisClient redisClient;

    @Inject
    ReactiveRedisClient reactiveRedisClient;

    public String get(String key) {
        Response response = redisClient.get(key);
        return response == null ? null : response.toString();
    }
}
```

When I request http://localhost:8080/redis/key1, I need to wait 5-10 seconds before there is a return value. If I remove the first three unavailable addresses configured by "quarkus.redis.hosts", I will get it soon To return value

### Expected behavior

Regardless of whether there are unavailable nodes, as long as the node where the key is located is available, the value should be returned immediately, which is very useful in the scenario of redis cluster shrinking (no need to restart the service)

### Actual behavior

When the unavailable node is configured in front of "quarkus.redis.hosts", the request response is slow. If configured in the back, it will not affect

### Output of `uname -a` or `ver`

Microsoft Windows [Version 10.0.22000.168]

### Output of `java -version`

java version "11.0.10" 2021-01-19

### Quarkus version or git rev

2.2.2.Final

### Build tool (ie. output of `mvnw --version` or `gradlew --version`)

Apache Maven 3.8.1

**Comment by machi1990** [comment]
Interesting, thanks for reporting @liuzhaozhao would you be able to come up with a complete reproducer, especially the redis cluster setup? Thanks

/cc @pmlopes would you have an idea of this behaviour?

**Comment by liuzhaozhao** [comment]
This is the redis server information


info Server
```
# Server
redis_version:5.0.5
redis_git_sha1:00000000
redis_git_dirty:0
redis_build_id:d843892b04e55eb8
redis_mode:cluster
os:Linux 4.4.0-116-generic x86_64
arch_bits:64
multiplexing_api:epoll
atomicvar_api:atomic-builtin
gcc_version:4.8.5
process_id:7559
run_id:1cf4cf7a95d34c9c22257e0f63822394768ae91f
tcp_port:7017
uptime_in_seconds:29629321
uptime_in_days:342
hz:10
configured_hz:10
lru_clock:4307859
executable:/opt/redis7017/src/redis-server
config_file:/opt/redis7017/redis7017.cnf
```

cluster info
```
cluster_state:ok
cluster_slots_assigned:16384
cluster_slots_ok:16384
cluster_slots_pfail:0
cluster_slots_fail:0
cluster_known_nodes:6
cluster_size:3
cluster_current_epoch:31
cluster_my_epoch:7
cluster_stats_messages_ping_sent:29508314
cluster_stats_messages_pong_sent:31960295
cluster_stats_messages_sent:61468609
cluster_stats_messages_ping_received:31960295
cluster_stats_messages_pong_received:29508291
cluster_stats_messages_received:61468586
```

cluster nodes
```
c3447bea53910b7eb301270ba708e2a177652e2a 127.0.0.1:7020@17020 slave 4df1c86d66f621c8b7c6de65ff768ca6f859cc38 1631697236029 1631697220000 25 connected
2db8dda0bde8d27fda515282b8ca16be8f3675d1 127.0.0.1:7019@17019 slave 8873c7dfaa93a2641152d5ee76a95f872ab11f61 0 1631697222035 19 connected
054ac3b72a19a62807205b019d869368fe2ee781 127.0.0.1:7021@17021 slave 20a7eb84db6d41dc6707e6dcf98afcc5b1cf25da 0 1631697221033 31 connected
20a7eb84db6d41dc6707e6dcf98afcc5b1cf25da 127.0.0.1:7016@17016 master - 0 1631697220029 1 connected 0-5460
4df1c86d66f621c8b7c6de65ff768ca6f859cc38 127.0.0.1:7017@17017 myself,master - 0 1631697221000 7 connected 5461-10922
8873c7dfaa93a2641152d5ee76a95f872ab11f61 127.0.0.1:7018@17018 master - 0 1631697223039 13 connected 10923-16383
```

Quarkus health information
The redis status is DOWN，but i can get key value
```json
{
    "status": "DOWN",
    "checks": [
        {
            "name": "SmallRye Reactive Messaging - liveness check",
            "status": "UP",
            "data": {
                "test-topic-event-in": "[OK]",
                "test-topic-in2": "[OK]",
                "test-topic-in": "[OK]",
                "test-topic-event-out": "[OK]",
                "test-topic-out": "[OK]",
                "test-topic-rabbitmq-consume": "[OK]",
                "test-topic-rabbitmq": "[OK]"
            }
        },
        {
            "name": "Database connections health check",
            "status": "UP"
        },
        {
            "name": "Reactive MySQL connections health check",
            "status": "UP",
            "data": {
                "": "up"
            }
        },
        {
            "name": "Redis connection health check",
            "status": "DOWN",
            "data": {
                "reason": "client [<default>]: io.vertx.core.impl.NoStackTraceThrowable: RedisClusterClient does not handle command ping, use non cluster client on the right node."
            }
        },
        {
            "name": "SmallRye Reactive Messaging - readiness check",
            "status": "UP",
            "data": {
                "test-topic-event-in": "[OK] - no subscription yet, so no connection to the Kafka broker yet",
                "test-topic-in2": "[OK]",
                "test-topic-in": "[OK]",
                "test-topic-event-out": "[OK]",
                "test-topic-out": "[OK]",
                "test-topic-rabbitmq-consume": "[OK]",
                "test-topic-rabbitmq": "[OK]"
            }
        },
        {
            "name": "SmallRye Reactive Messaging - startup check",
            "status": "UP",
            "data": {
                "test-topic-event-in": "[OK] - no subscription yet, so no connection to the Kafka broker yet",
                "test-topic-in2": "[OK]",
                "test-topic-in": "[OK]",
                "test-topic-event-out": "[OK]",
                "test-topic-out": "[OK]"
            }
        }
    ]
}
```

**Comment by liuzhaozhao** [comment]
You can try to add some non-existent node addresses before the quarkus.redis.hosts configuration value of a redis cluster to simulate the problem @machi1990

**Comment by pmlopes** [comment]
@liuzhaozhao @machi1990 I've tried to reproduce the issue in a unit test (upstream at the vertx-redis-client):

```java
  @Test(timeout = 30_000)
  public void testConnectTime(TestContext should) {
    final Async test = should.async();

    RedisOptions options = new RedisOptions()
      .setType(RedisClientType.CLUSTER)
      .setUseReplicas(RedisReplicas.SHARE)
      // the first hosts are fake to simulate wrong setup
      .addConnectionString("redis://127.0.0.1:7005")
      .setMaxPoolSize(8)
      .setMaxPoolWaiting(16);

    long t0 = System.currentTimeMillis();
    client = Redis.createClient(rule.vertx(), options);
    long t1 = System.currentTimeMillis();

    client
      .connect(onCreate -> {
        long t2 = System.currentTimeMillis();

        should.assertTrue(onCreate.succeeded());

        System.out.println(t1 - t0);
        System.out.println(t2 - t1);
        test.complete();
      });
  }
```

The cluster is using the ports 7000 to 7005

I see the following output:

```
0
214
```

So the baseline is using a cold jvm (test runs during a maven test phase) find and connect to all nodes 214ms

If I add fake entries:

```java
  @Test(timeout = 30_000)
  public void testConnectTime(TestContext should) {
    final Async test = should.async();

    RedisOptions options = new RedisOptions()
      .setType(RedisClientType.CLUSTER)
      .setUseReplicas(RedisReplicas.SHARE)
      // the first hosts are fake to simulate wrong setup
      .addConnectionString("redis://127.0.0.1:9999")
      .addConnectionString("redis://127.0.0.1:8888")
      .addConnectionString("redis://127.0.0.1:7005")
      .setMaxPoolSize(8)
      .setMaxPoolWaiting(16);

    long t0 = System.currentTimeMillis();
    client = Redis.createClient(rule.vertx(), options);
    long t1 = System.currentTimeMillis();

    client
      .connect(onCreate -> {
        long t2 = System.currentTimeMillis();

        should.assertTrue(onCreate.succeeded());

        System.out.println(t1 - t0);
        System.out.println(t2 - t1);
        test.complete();
      });
  }
```

I see:

```
0
216
```

Which gives almost the same result, so if you're observing a delay in the magnitude of seconds I think the issue is not the availability of the hosts.

I can imagine a situation where the current cluster code can be slow. That is when we don't use pools and are constantly reconnecting, given that the current client expects to have a working connection to each slot is can take some time. An improvement that can be a nice open source contribution is to attempt to make the connection to the clients lazy, as some users, may just want to send a command and quit.

**Comment by liuzhaozhao** [comment]
@pmlopes @machi1990 I tried to build a new redis cluster, but this problem still exists.
By tracking the code, I found the reason for the slow request. The code execution order of redis request is as follows:

1. io.quarkus.redis.client.runtime.RedisClientImpl.get(String arg0)
2. io.vertx.redis.client.impl.RedisAPIImpl.send(Command cmd, String... args)
3. io.vertx.redis.client.impl.BaseRedisClient.send(Request command)
4. io.vertx.redis.client.impl.RedisClusterClient.connect(List<String> endpoints, int index, Handler<AsyncResult<RedisConnection>> onConnect)

The first parameter of step 4 contains all the configured redis node addresses (including unavailable ones), and the connect method will get the first available endpoint according to the endpoints order to load the slots, so every redis request will try and Unavailable nodes to establish a connection, resulting in longer time-consuming

Is it necessary to add a check function to temporarily remove unavailable nodes, wait for a period and then try to join, and remove them permanently when the maximum number of attempts is reached?
